File: src/main/java/me/lokka30/levelledmobs/customdrops/CustomDropInstance.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.customdrops;

import java.util.LinkedList;
import java.util.List;
import me.lokka30.levelledmobs.misc.CustomUniversalGroups;
import org.bukkit.entity.EntityType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Holds a mob or group instance and associates it with a list of custom drop items. This is where
 * the override for a mob / group is set
 *
 * @author stumper66
 * @since 2.4.0
 */
public class CustomDropInstance {

    public CustomDropInstance(final @Nullable EntityType associatedMob) {
        this.associatedMob = associatedMob;
        this.entityGroup = null;
        this.customItems = new LinkedList<>();
        this.overallPermissions = new LinkedList<>();
        this.isBabyMob = false;
    }

    public CustomDropInstance(final @NotNull EntityType associatedMob, final boolean isBabyMob) {
        this.associatedMob = associatedMob;
        this.entityGroup = null;
        this.customItems = new LinkedList<>();
        this.overallPermissions = new LinkedList<>();
        this.isBabyMob = isBabyMob;
    }

    public CustomDropInstance(final @NotNull CustomUniversalGroups entityGroup) {
        this.associatedMob = null;
        this.entityGroup = entityGroup;
        this.customItems = new LinkedList<>();
        this.overallPermissions = new LinkedList<>();
        isBabyMob = false;
    }

    private final @Nullable EntityType associatedMob;
    private final @Nullable CustomUniversalGroups entityGroup;
    public final @NotNull List<CustomDropBase> customItems;
    public Float overallChance;
    public final @NotNull List<String> overallPermissions;
    public Boolean overrideStockDrops;
    boolean utilizesGroupIds;
    final boolean isBabyMob;

    public void combineDrop(final CustomDropInstance dropInstance) {
        if (dropInstance == null) {
            throw new NullPointerException("dropInstance");
        }

        this.overrideStockDrops = dropInstance.overrideStockDrops;

        if (dropInstance.utilizesGroupIds) {
            this.utilizesGroupIds = true;
        }

        this.customItems.addAll(dropInstance.customItems);
    }

    public @NotNull String getMobOrGroupName() {
        if (this.associatedMob != null) {
            return this.associatedMob.name();
        } else if (this.entityGroup != null) {
            return this.entityGroup.name();
        } else {
            return ""; // this return should never happen
        }
    }

    public @Nullable EntityType getAssociatedMobType(){
        return this.associatedMob;
    }

    public boolean getOverrideStockDrops(){
        return this.overrideStockDrops != null && this.overrideStockDrops;
    }

    public @NotNull String toString() {
        if (this.associatedMob != null) {
            return this.getOverrideStockDrops() ?
                this.associatedMob.name() + " - override" :
                this.associatedMob.name();
        } else if (this.entityGroup != null) {
            return this.getOverrideStockDrops() ?
                this.entityGroup + " - override" :
                this.entityGroup.toString();
        } else {
            return "CustomDropInstance";
        }
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/customdrops/CustomDropInstance.kt
package io.github.arcaneplugins.levelledmobs.customdrops

import io.github.arcaneplugins.levelledmobs.misc.CustomUniversalGroups
import org.bukkit.entity.EntityType

/**
 * Holds a mob or group instance and associates it with a list of custom drop items. This is where
 * the override for a mob / group is set
 *
 * @author stumper66
 * @since 2.4.0
 */
class CustomDropInstance {
    var associatedMob: EntityType? = null
        private set
    private var entityGroup: CustomUniversalGroups? = null
    var customItems = mutableListOf<CustomDropBase>()
    var overallChance: Float? = null
    var overallPermissions = mutableListOf<String>()
    var overrideStockDrops: Boolean? = null
    var utilizesGroupIds: Boolean = false
    var isBabyMob: Boolean = false

    constructor(associatedMob: EntityType?){
        this.associatedMob = associatedMob
    }

    constructor(associatedMob: EntityType?, isBabyMob: Boolean){
        this.associatedMob = associatedMob
        this.isBabyMob = isBabyMob
    }

    constructor(entityGroup: CustomUniversalGroups){
        this.entityGroup = entityGroup
    }

    fun combineDrop(dropInstance: CustomDropInstance?) {
        if (dropInstance == null) {
            throw NullPointerException("dropInstance")
        }

        this.overrideStockDrops = dropInstance.overrideStockDrops

        if (dropInstance.utilizesGroupIds) {
            this.utilizesGroupIds = true
        }

        customItems.addAll(dropInstance.customItems)
    }

    fun getMobOrGroupName(): String {
        return if (this.associatedMob != null) {
            associatedMob!!.name
        } else if (this.entityGroup != null) {
            entityGroup!!.name
        } else {
            "" // this return should never happen
        }
    }

    val getOverrideStockDrops: Boolean
        get() = this.overrideStockDrops != null && this.overrideStockDrops!!

    override fun toString(): String {
        return if (this.associatedMob != null) {
            if (getOverrideStockDrops) associatedMob!!.name + " - override" else associatedMob!!.name
        } else if (this.entityGroup != null) {
            if (getOverrideStockDrops) entityGroup.toString() + " - override" else entityGroup.toString()
        } else {
            "CustomDropInstance"
        }
    }
}